Add blog post: Scaling Data Pipelines for a Growth-Stage Fintech with Incremental Models

Adedamola-Onabanjo opened the pull request:
## What are you changing in this pull request and why?

This PR adds a new blog post titled **"Scaling Data Pipelines for a Growth-Stage Fintech with Incremental Models"**.

The article walks through Kuda’s real-world implementation of `incremental` dbt models using BigQuery, and how they enabled scalable, cost-effective analytics in a high-growth environment.

It covers:
- Append, insert_overwrite, and merge strategies
- Performance tuning and partitioning techniques
- Real-world cost/time savings
- Inline `jinja` and `sql` code snippets
- Production lessons learned

This post was prepared and refined with guidance from the dbt Labs Developer Relations team.

## Checklist
- [x] The changes in this PR meet the [docs style guide/fundamentals](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/content-style-guide.md) required for all content.
- [x] Applied the proper versioning rules if applicable.
- [x] Article includes clean front matter, formatted markdown tables, fenced code blocks, and inline tags.
- [x] The blog post has been previewed locally and renders cleanly.

Adedamola-Onabanjo committed:
Add blog post: Scaling Data Pipelines for a Growth-Stage Fintech with Incremental Models

Initial version of the blog post based on the final draft shared with Jason. Includes full MDX formatting and front matter

Adedamola-Onabanjo committed:
Update 2024-04-25-scaling-data-pipelines-kuda.mdx

runleonarun commented:
Hi @Adedamola-Onabanjo The Docs team will get this reviewed and posted for you! It sounds liek it got a technical review from @SamHarting

runleonarun committed:
Update 2024-04-25-scaling-data-pipelines-kuda.mdx

Removing ampersand

mirnawong1 committed:
Merge branch 'current' into current

mirnawong1 committed:
Update 2024-04-25-scaling-data-pipelines-kuda.mdx

fix frontmatter and backticks

mirnawong1 committed:
Update 2024-04-25-scaling-data-pipelines-kuda.mdx

frontmatter

mirnawong1 commented:
build error due to author not found in authors.yml. 
` [cause]: Error: Blog author with key "adedamola_onabanjo" not found in the authors map file.`

mirnawong1 committed:
Merge branch 'current' into current

Adedamola-Onabanjo committed:
Add files via upload

Adedamola-Onabanjo committed:
Update authors.yml

Add author profile and image for blog post

Adedamola-Onabanjo commented:
Hi @mirnawong1

I’ve uploaded my profile picture and updated the authors.yml file as requested.
All changes have been committed. Please let me know if there’s anything else needed!

Thank you again for the guidance.

mirnawong1 commented on website/blog/authors.yml:
```diff
@@ -1,4 +1,12 @@
 ---
+adedamola_onabanjo:
+  description: Adedamola builds scalable BI and analytics solutions for fintech companies, passionate about empowering data-driven decisions.
+  image_url: /img/blog/authors/adedamola_onabanjo.png
```
```suggestion
  image_url: /img/blog/authors/adedamola-onabanjo.png
```

mirnawong1 committed:
Update website/blog/authors.yml

mirnawong1 committed:
Merge branch 'current' into current

mirnawong1 commented on website/blog/authors.yml:
```diff
@@ -1,4 +1,12 @@
 ---
+adedamola_onabanjo:
+  description: Adedamola builds scalable BI and analytics solutions for fintech companies, passionate about empowering data-driven decisions.
+  image_url: /img/blog/authors/adedamola-onabanjo.png
```
```suggestion
  image_url: /img/blog/adedamola-onabanjo.png
```

mirnawong1 committed:
Update website/blog/authors.yml

mirnawong1 commented on website/blog/authors.yml:
```diff
@@ -1,4 +1,12 @@
 ---
+adedamola_onabanjo:
+  description: Adedamola builds scalable BI and analytics solutions for fintech companies, passionate about empowering data-driven decisions.
+  image_url: /img/blog/adedamola-onabanjo.png
```
changed url since image wasn't displaying properly

mirnawong1 replied:
https://docs-getdbt-2c9t1b2ve-dbt-labs.vercel.app/blog

mirnawong1 commented on website/blog/2024-04-25-scaling-data-pipelines-kuda.mdx:
```diff
@@ -0,0 +1,254 @@
+
+## Introduction
+
+Building scalable data pipelines in a fast-growing fintech can feel like fixing a bike while riding it. You must keep insights flowing even as data volumes explode. At Kuda (a Nigerian neo-bank), we faced this problem as our user base surged. Traditional batch ETL (rebuilding entire tables each run) started to buckle; pipelines took hours, and costs ballooned. We needed to keep data fresh without ...
```
```suggestion
Building scalable data pipelines in a fast-growing fintech can feel like fixing a bike while riding it. You must keep insights flowing even as data volumes explode. At Kuda (a Nigerian neo-bank), we faced this problem as our user base surged. Traditional batch ETL (rebuilding entire tables each run) started to buckle; pipelines took hours, and costs ballooned. We needed to keep data fresh without reprocessing everything. Our solution was to leverage dbt’s [incremental models](/docs/build/incremental-models), which process only new or changed records. This dramatically cut run times and curbed our BigQuery costs, letting us scale efficiently.
```

mirnawong1 commented on website/blog/2024-04-25-scaling-data-pipelines-kuda.mdx:
```diff
@@ -0,0 +1,254 @@
+
+## Approach: Incremental Models & Key Strategies
+
+We tackled these issues by embracing dbt’s incremental models, which process only new or updated records since the last run. Instead of monolithic daily rebuilds, our models continuously ingested changes in small bites. Below, we outline our key incremental strategies—`append`, `insert_overwrite`, and `merge`—and how we tuned performance and cost.
```
```suggestion
We tackled these issues by embracing dbt’s incremental models, which process only new or updated records since the last run. Instead of monolithic daily rebuilds, our models continuously ingested changes in small bites. Below, we outline our key [incremental strategies](/docs/build/incremental-strategy) —`append`, `insert_overwrite`, and `merge` — and how we tuned performance and cost.
```

mirnawong1 commented on website/blog/2024-04-25-scaling-data-pipelines-kuda.mdx:
```diff
@@ -0,0 +1,254 @@
+
+This is the simplest incremental approach: Each run adds new rows to the existing table and never touches old rows. It's ideal for append-only data (e.g. logs or transactions that never change after insertion).
+
+In dbt, using append is straightforward. We configure the model as incremental and specify `incremental_strategy='append'` (for BigQuery and similar warehouses). In the SQL, we filter the source to only new records since the last load. For example, to incrementally load new transactions:
```
```suggestion
In dbt, using append is straightforward. We configure the model as incremental and specify `incremental_strategy='append'` (for [BigQuery](/reference/resource-configs/bigquery-configs#merge-behavior-incremental-models) and [similar warehouses](/docs/build/incremental-strategy#supported-incremental-strategies-by-adapter)). In the SQL, we filter the source to only new records since the last load. For example, to incrementally load new transactions:
```

mirnawong1 replied:
hey @Adedamola-Onabanjo , [it looks like ](https://docs.getdbt.com/docs/build/incremental-strategy#supported-incremental-strategies-by-adapter) the `append` strategy isn’t supported in BQ? ([core code here](https://github.com/dbt-labs/dbt-adapters/blob/7a568afe705d8e5884eef7416f0e7818b01aba84/dbt-bigquery/src/dbt/include/bigquery/macros/materializations/incremental.sql) -- do you have it configured and working?

Adedamola-Onabanjo replied:
Thanks for flagging this, @mirnawong1 
I had intended the append reference more generally (Snowflake) and not BigQuery, which is the warehouse I work with.
Happy to update the wording for clarity. Please let me know if I should do a bulk edit (with the other formatting errors) now or wait till you are done reviewing

mirnawong1 replied:
hey @Adedamola-Onabanjo , awesome thank you for clarifying! i've updated the bold bits and done with my review -- happy for you to go ahead!

mirnawong1 replied:
hey @Adedamola-Onabanjo !  just wanted to check in on this to see if you've made those updates?

mirnawong1 commented on website/blog/2024-04-25-scaling-data-pipelines-kuda.mdx:
```diff
@@ -0,0 +1,254 @@
+| 2023-10-01 | 12001 | C021 | ₦8,000 | pending |
+| 2023-10-01 | 12002 | C022 | ₦4,200 | completed |
+
+_(Partition to be overwritten highlighted in bold.)_
```
hey @Adedamola-Onabanjo , just a heads up that the partitioned data isn't highlighted in bold - should it be?

Adedamola-Onabanjo replied:
@mirnawong1 yes please. i skipped that in my md edit. can I go ahead and make it bold?

mirnawong1 commented on website/blog/2024-04-25-scaling-data-pipelines-kuda.mdx:
```diff
@@ -0,0 +1,254 @@
+| 2023-10-01 | 12004 | C024 | ₦3,900 | completed |
+| 2023-10-01 | 12005 | C025 | ₦5,500 | completed |
+
+_(New partition data is shown in bold, replacing the old partition.)_
```
same here about bolding the new data

mirnawong1 replied:
presume it's 12003-12005?

Adedamola-Onabanjo replied:
@mirnawong1 same here too

Adedamola-Onabanjo replied:
> presume it's 12003-12005?

correct @mirnawong1

mirnawong1 committed:
Update website/blog/2024-04-25-scaling-data-pipelines-kuda.mdx

mirnawong1 committed:
Update website/blog/2024-04-25-scaling-data-pipelines-kuda.mdx

mirnawong1 committed:
Update 2024-04-25-scaling-data-pipelines-kuda.mdx

adding bolded info

mirnawong1 committed:
Update website/blog/2024-04-25-scaling-data-pipelines-kuda.mdx

mirnawong1 committed:
Merge branch 'current' into current

mirnawong1 committed:
Merge branch 'current' into current

Adedamola-Onabanjo committed:
Update 2024-04-25-scaling-data-pipelines-kuda.mdx

Clarify adapter support for append incremental strategy and change format to bold where necessary

mirnawong1 committed:
Merge branch 'current' into current

Adedamola-Onabanjo committed:
Update 2024-04-25-scaling-data-pipelines-kuda.mdx

mirnawong1 commented on website/blog/2024-04-25-scaling-data-pipelines-kuda.mdx:
```diff
@@ -0,0 +1,258 @@
+authors: [adedamola_onabanjo]
+tags: [analytics, dbt Cloud, fintech, BigQuery]
+hide_table_of_contents: false
+date: 2025-04-23
```
```suggestion
date: 2025-05-01
```

mirnawong1 committed:
Update website/blog/2024-04-25-scaling-data-pipelines-kuda.mdx
